src/processor/operator/copy/copy_node.cpp

```diff
@@ -54,171 +50,152 @@ CopyNode::CopyNode(std::shared_ptr<CopyNodeSharedState> sharedState, CopyNodeInf
+                sliceDataChunk(*resultSet->getDataChunk(0), copyNodeInfo.dataColumnPoses,
+                    (int64_t)numAppendedTuplesInNodeGroup);
+            }
         }
-        flushChunksAndPopulatePKIndex(
-            columnChunks, startRowIdx, endRowIdx, filePath, startRowIdxInFile);
+    }
+    if (localNodeGroup->getNumNodes() > 0) {
+        sharedState->appendLocalNodeGroup(std::move(localNodeGroup));
     }
 }
 
-void CopyNode::finalize(kuzu::processor::ExecutionContext* context) {
-    auto tableID = copyNodeInfo.table->getTableID();
-    if (sharedState->pkIndex) {
-        sharedState->pkIndex->flush();
-    }
-    for (auto& column : sharedState->columns) {
-        column->saveToFile();
-    }
-    for (auto& relTableSchema :
-        copyNodeInfo.catalog->getAllRelTableSchemasContainBoundTable(tableID)) {
-        copyNodeInfo.relsStore->getRelTable(relTableSchema->tableID)
-            ->batchInitEmptyRelsForNewNodes(relTableSchema, sharedState->numRows);
+void CopyNode::sliceDataChunk(
+    const DataChunk& dataChunk, const std::vector<DataPos>& dataColumnPoses, offset_t offset) {
+    for (auto& dataColumnPos : dataColumnPoses) {
+        assert(dataColumnPos.dataChunkPos == 0);
+        ArrowColumnVector::slice(
+            dataChunk.valueVectors[dataColumnPos.valueVectorPos].get(), offset);
     }
-    copyNodeInfo.table->getNodeStatisticsAndDeletedIDs()->setNumTuplesForTable(
-        tableID, sharedState->numRows);
-    auto outputMsg = StringUtils::string_format("{} number of tuples has been copied to table: {}.",
-        sharedState->numRows,
-        copyNodeInfo.catalog->getReadOnlyVersion()->getTableName(tableID).c_str());
-    FactorizedTableUtils::appendStringToTable(
-        sharedState->table.get(), outputMsg, context->memoryManager);
 }
 
-void CopyNode::flushChunksAndPopulatePKIndex(
-    const std::vector<std::unique_ptr<InMemColumnChunk>>& columnChunks, offset_t startNodeOffset,
-    offset_t endNodeOffset, const std::string& filePath, row_idx_t startRowIdxInFile) {
-    // Flush each page within the [StartOffset, endOffset] range.
-    for (auto i = 0u; i < sharedState->columns.size(); i++) {
-        sharedState->columns[i]->flushChunk(columnChunks[i].get());
-    }
-    if (sharedState->pkIndex) {
-        // Populate the primary key index.
-        populatePKIndex(columnChunks[sharedState->pkColumnID].get(),
-            sharedState->columns[sharedState->pkColumnID]->getInMemOverflowFile(), startNodeOffset,
-            (endNodeOffset - startNodeOffset + 1), filePath, startRowIdxInFile);
-    }
+void CopyNode::writeAndResetNodeGroup(node_group_idx_t nodeGroupIdx,
+    PrimaryKeyIndexBuilder* pkIndex, column_id_t pkColumnID, NodeTable* table,
+    NodeGroup* nodeGroup) {
+    nodeGroup->setNodeGroupIdx(nodeGroupIdx);
+    auto startOffset = StorageUtils::getStartOffsetForNodeGroup(nodeGroupIdx);
+    auto numNodes = nodeGroup->getNumNodes();
```

Comment:
You can put this inside the if and inline when you call `populatePKIndex`.

Reply:
changed.